File: tests/c++/test_geo_coordinate_bounding_box2d.cpp
#include <gtest/gtest.h>

#include "quad_key/geo_coordinate_bounding_box2d.h"
#include "quad_key/geo_coordinate2d.h"

#include <iostream>

namespace quad_keys
{
    class test_geo_coordinate_bounding_box2d : public ::testing::Test
    {
    protected:
      virtual void SetUp() {}
      virtual void TearDown() {}
    };

    TEST(GEOCOORDINATEBOUNDINGBOX2D, default_ctor)
    {
      geo_coordinate_bounding_box2d empty;
      ASSERT_FALSE(empty.is_valid());
      ASSERT_FALSE(empty.contains(
        geo_coordinate2d(longitude(0.0), latitude(0.0))));
    }

    TEST(GEOCOORDINATEBOUNDINGBOX2D, ctor_standard_construction)
    {
      { // Lower Left / Upper Right Construction
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_coordinate2d(longitude(180.0), latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(90.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }

      { // Upper Left / Lower Right Construction
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(0.0), latitude(90.0)),
          geo_coordinate2d(longitude(180.0), latitude(0.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(90.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }

      { // Upper Right / Lower Left Construction
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(180.0), latitude(0.0)),
          geo_coordinate2d(longitude(0.0), latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(90.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }

      { // Lower Right / Upper Left Construction
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(180.0), latitude(0.0)),
          geo_coordinate2d(longitude(0.0), latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(90.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }
    }

    TEST(GEOCOORDINATEBOUNDINGBOX2D, ctor_linear_box_and_point)
    {
      { // Horizontal
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_coordinate2d(longitude(0.0), latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(90.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }

      { // Vertical
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_coordinate2d(longitude(180.0), latitude(0.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(0.0)),
          geo_box.upper_right());
        EXPECT_TRUE(geo_box.is_valid());
      }

      { // Point
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(0.0), latitude(0.0)),
          geo_coordinate2d(longitude(0.0), latitude(0.0)));

        EXPECT_TRUE(geo_box.is_valid());
      }
    }

    TEST(GEOCOORDINATEBOUNDINGBOX2D, contains_or_not)
    {
      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)));

        EXPECT_TRUE(geo_box.is_valid());

        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(0.0),
          latitude(0.0))));

        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(90.0),
          latitude(45.0))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(90.0),
          latitude(-45.0))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(-90.0),
          latitude(-45.0))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(-90.0),
          latitude(45.0))));

        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(90.01),
          latitude(45.01))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(90.01),
          latitude(-45.01))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(-90.01),
          latitude(-45.01))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(-90.01),
          latitude(45.01))));
      }

      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_coordinate2d(longitude(0.1), latitude(0.1)));

        EXPECT_EQ(geo_coordinate2d(longitude(0.1), latitude(0.1)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        EXPECT_TRUE(geo_box.is_valid());

        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(90.0),
          latitude(45.0))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(90.0),
          latitude(0.1))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(0.1),
          latitude(0.1))));
        EXPECT_TRUE(geo_box.contains(geo_coordinate2d(longitude(0.1),
          latitude(45.0))));

        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(90.01),
          latitude(45.01))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(90.01),
          latitude(0.0))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(0.0),
          latitude(0.0))));
        EXPECT_FALSE(geo_box.contains(geo_coordinate2d(longitude(0.0),
          latitude(45.01))));
      }
    }

    TEST(GEOCOORDINATEBOUNDINGBOX2D, grow_to_contain)
    {
      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_coordinate2d(longitude(90.0), latitude(45.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        geo_box.grow_to_contain(geo_coordinate2d(longitude(180.0),
          latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(90.0)),
          geo_box.upper_right());
      }

      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_coordinate2d(longitude(90.0), latitude(45.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        geo_box.grow_to_contain(geo_coordinate2d(longitude(-180.0),
          latitude(-90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-180.0), latitude(-90.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());
      }

      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_coordinate2d(longitude(90.0), latitude(45.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        geo_box.grow_to_contain(geo_coordinate2d(longitude(-180.0),
          latitude(90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-180.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(90.0)),
          geo_box.upper_right());
      }

      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_coordinate2d(longitude(90.0), latitude(45.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        geo_box.grow_to_contain(geo_coordinate2d(longitude(180.0),
          latitude(-90.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-90.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(180.0), latitude(45.0)),
          geo_box.upper_right());
      }

      {
        geo_coordinate_bounding_box2d geo_box(
          geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_coordinate2d(longitude(90.0), latitude(45.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());

        geo_box.grow_to_contain(geo_coordinate2d(longitude(0.0),
          latitude(0.0)));

        EXPECT_EQ(geo_coordinate2d(longitude(-90.0), latitude(-45.0)),
          geo_box.lower_left());
        EXPECT_EQ(geo_coordinate2d(longitude(90.0), latitude(45.0)),
          geo_box.upper_right());
      }
    }
}

File: bindings/python/quad_keys/tests/test_geo_coordinate_bounding_box2d.py
"""
Unit Tests for GeoCoordinateBoundingBox2d
"""

from __future__ import print_function

import unittest

from quad_keys.quad_keys import GeoCoordinateBoundingBox2d
from quad_keys.quad_keys import GeoCoordinate2d
from quad_keys.quad_keys import Longitude
from quad_keys.quad_keys import Latitude


class TestGeoCoordinateBoundingBox2d(unittest.TestCase):
    """Class for running unit tests."""

    def test_default_ctor(self):
        """default ctor construction."""
        empty = GeoCoordinateBoundingBox2d()
        self.assertFalse(empty.is_valid())
        self.assertFalse(empty.contains(GeoCoordinate2d(Longitude(0.0),
                                                        Latitude(0.0))))

    def test_ctor_standard_construction(self):
        """default ctor construction."""
        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(180.0), Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(0.0), Latitude(90.0)),
            GeoCoordinate2d(Longitude(180.0), Latitude(0.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(0.0), Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(0.0), Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

    def test_ctor_linear_box_and_point(self):
        """default ctor construction."""
        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(0.0), Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(90.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(180.0), Latitude(0.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
                         geobox.upper_right())
        self.assertTrue(geobox.is_valid())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
            GeoCoordinate2d(Longitude(0.0), Latitude(0.0)))
        self.assertTrue(geobox.is_valid())

    def test_contains_or_not(self):
        """containment test."""
        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)))
        self.assertTrue(geobox.is_valid())
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(0.0),
                                                        Latitude(0.0))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(90.0),
                                                        Latitude(45.0))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(90.0),
                                                        Latitude(-45.0))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(-90.0),
                                                        Latitude(-45.0))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(-90.0),
                                                        Latitude(45.0))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(90.01),
                                                         Latitude(45.01))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(90.01),
                                                         Latitude(-45.01))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(-90.01),
                                                         Latitude(-45.01))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(-90.01),
                                                         Latitude(45.01))))

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
            GeoCoordinate2d(Longitude(0.1), Latitude(0.1)))
        self.assertEqual(GeoCoordinate2d(Longitude(0.1), Latitude(0.1)),
                         geobox.lower_left())
        self.assertTrue(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                        geobox.upper_right())
        self.assertTrue(geobox.is_valid())
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(90.0),
                                                        Latitude(45.0))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(90.0),
                                                        Latitude(0.1))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(0.1),
                                                        Latitude(0.1))))
        self.assertTrue(geobox.contains(GeoCoordinate2d(Longitude(0.1),
                                                        Latitude(45.0))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(90.01),
                                                         Latitude(45.01))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(90.01),
                                                         Latitude(0.0))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(0.0),
                                                         Latitude(0.0))))
        self.assertFalse(geobox.contains(GeoCoordinate2d(Longitude(0.0),
                                                         Latitude(45.01))))

    def test_grow_to_contain(self):
        """grow to contain test"""
        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
        geobox.grow_to_contain(GeoCoordinate2d(Longitude(180.0),
                                               Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
                         geobox.upper_right())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
        geobox.grow_to_contain(
            GeoCoordinate2d(Longitude(-180.0), Latitude(-90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-180.0),
                                         Latitude(-90.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)))
        self.assertTrue(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
                        geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
        geobox.grow_to_contain(GeoCoordinate2d(Longitude(-180.0),
                                               Latitude(90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-180.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(90.0)),
                         geobox.upper_right())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
        geobox.grow_to_contain(GeoCoordinate2d(Longitude(180.0),
                                               Latitude(-90.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-90.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(180.0),
                                         Latitude(45.0)),
                         geobox.upper_right())

        geobox = GeoCoordinateBoundingBox2d(
            GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
            GeoCoordinate2d(Longitude(90.0), Latitude(45.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
        geobox.grow_to_contain(GeoCoordinate2d(Longitude(0.0),
                                               Latitude(0.0)))
        self.assertEqual(GeoCoordinate2d(Longitude(-90.0),
                                         Latitude(-45.0)),
                         geobox.lower_left())
        self.assertEqual(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
                         geobox.upper_right())
